File: Projeto Didio Pizza/Didio Pizza Node/backend/src/service/pagamentoService.js
import * as pagamentoRepository from '../repository/pagamentoRepository.js';
import connection from '../repository/connection.js';
import * as pedidoRepository from '../repository/pedidoRepository.js';


export async function realizarPagamento(clienteId, pedidoId, tipo) {

    if (!pedidoId || typeof tipo !== 'string') {
        throw new Error('Pedido e tipo são obrigatórios.');
    }

    tipo = String(tipo).toUpperCase();

    if (!['DINHEIRO', 'CARTAO', 'PIX'].includes(tipo)) {
        throw new Error('Tipo inválido.');
    }

    let pedido = await pedidoRepository.buscarPedidoPorId(pedidoId, clienteId);

    if (!pedido) {
        throw new Error('Pedido não encontrado.');
    }

    if (pedido.status !== 'REGISTRADO') {
        throw new Error('Pedido não está disponível para pagamento.');
    }

    if (pedido.total <= 0) {
        throw new Error('Valor do pedido inválido.');
    }

    let pagamentoExistente = await pagamentoRepository.buscarPagamentoPorPedido(pedidoId);

    if (pagamentoExistente) {
        throw new Error('Pedido já pago.');
    }

    const conn = await connection.getConnection();

    try {
        await conn.beginTransaction();

        let pagamentoId = await pagamentoRepository.inserirPagamento(conn, clienteId, pedidoId, tipo, pedido.total);

        let linhas = await pedidoRepository.atualizarStatus(conn, pedidoId, 'PREPARANDO');

        if (linhas === 0) {
            throw new Error('Erro ao atualizar status do pedido.');
        }

        await pedidoRepository.inserirHistorico(conn, pedidoId, 'PREPARANDO');

        await conn.commit();

        return pagamentoId;

    } catch (err) {
        await conn.rollback();
        throw err;
    } finally {
        conn.release();
    }
}

File: Projeto Didio Pizza/Didio Pizza Springboot/backend/src/main/java/com/pizzaria/didiopizza/service/PagamentoService.java
package com.pizzaria.didiopizza.service;

import java.util.List;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.pizzaria.didiopizza.repository.PagamentoRepository;
import com.pizzaria.didiopizza.repository.PedidoRepository;

@Service
public class PagamentoService {
    private final PagamentoRepository pagamento;
    private final PedidoRepository pedido;

    public PagamentoService(PagamentoRepository pagamento, PedidoRepository pedido) {
        this.pagamento = pagamento;
        this.pedido = pedido;
    }

    @Transactional
    public Integer realizarPagamento(Integer clienteId, Integer pedidoId, String tipo) {
        if (pedidoId == null || tipo == null)
            throw new RuntimeException("Pedido e tipo são obrigatórios.");

        tipo = tipo.toUpperCase();
        if (!List.of("DINHEIRO", "CARTAO", "PIX").contains(tipo))
            throw new RuntimeException("Tipo inválido.");

        var p = pedido.buscarPedidoPorId(pedidoId, clienteId);
        if (p == null)
            throw new RuntimeException("Pedido não encontrado.");

        if (!"REGISTRADO".equals(String.valueOf(p.get("status"))))
            throw new RuntimeException("Pedido não está disponível para pagamento.");

        if (Double.parseDouble(String.valueOf(p.get("total"))) <= 0)
            throw new RuntimeException("Valor do pedido inválido.");

        if (pagamento.buscarPagamentoPorPedido(pedidoId) != null)
            throw new RuntimeException("Pedido já pago.");

        Integer id = pagamento.inserirPagamento(clienteId, pedidoId, tipo, p.get("total"));
        if (pedido.atualizarStatus(pedidoId, "PREPARANDO") == 0)
            throw new RuntimeException("Erro ao atualizar status do pedido.");

        pedido.inserirHistorico(pedidoId, "PREPARANDO");

        return id;
    }
}
